### DISCUSSION TITLE
ACL on cmnd topic

### DISCUSSION THREAD

**Discussion by ouinouin** [arendst/Tasmota#14707] [Ideas]
hi, 
as a part of an associative french ISP (http://tetaneutral.net/), we started a small project to encourage people to have some "opensource" IOT devices specially for power monitoring (more especially to monitor the fridges as they are subject to a huge change of power consumption depending on their age/health).
we intend to give monitoring plugs to participants of this project, and , each plug will connect to our infrastructure (one server on the internet with nodered, influxdb , and grafana.)
we want people to be able to compare their data on the server, so they dont need any additional hardware ath home.
for this we use mqtt over tls.
the problem is to blcok the downstream commands, for sure we could block the topics on the mqtt server, but for us , this is not enough since the user of the plug should have the control of his material, and the security of the fridge supply shouldnt depend on our server (imagine someone breaking the server and switching off all the fridges of the contest remotely.
so i launch this discusssion to start some reflexion on some kind of ACL on the tasmota side, or at least some function to disable or enable the subscription on the whole CMND topic.
as a low skilled developper , i ll first give basic trials, but dont hesitate to discuss if you re interrested, we might get some skilled people from our association on this subject.
thanks for feedback.

**Comment by barbudor** [comment]
Bonjour

I don't know why you would exclude managing the ACL on the MQTT Server. This sounds the most convenient and proper way from my point of view.
With Moquitto for exemple you can control who is accessing topics. 
So you could assign a user/password to each of your participant so they can connect to the MQTT broker, but each would only be able to publish to the topics of it's own equipment.
Same, you could have an user/password for nodered which has the right to subscribe to any data (tele/# for exemple) but prevent other users to directly "spy" on the data of the others.

ACL on Tasmota doesn't make much sense as MQTT protocol doesn't provide any means for the subscriber (Tasmota to its own cmnd/ topic) to identify from who the command is comming.

Happy to give you more hints on Mosquitto ACL is needed.

**Reply by ouinouin** [reply]
@barbudor Bonjour aussi :-)

By ACL I would more mean , allow or disallow the use of each command through mqtt, not by user but more having a list to allow or disallow each command (or disallow (meaning technically not subscribing to CMND) all the commands).
technically, i m ok to do this on server side, bu this implies living with the risk of having your appliance shut down if somebody is doing something stupid on the server. if the "access control" is done on the tasmota side, then you would sleep better not asking yourself if you ll find your fridge off because of an attack on a server somewhere on the internet :-).

**Reply by barbudor** [reply]
The best would be that each participant has a MQTT broker of their own on their home with a rule to forward only the SENSOR messages to the aggregating MQTT broker.
Definitively standard MQTT systems is not designed for multi-tenant. Sharing a single MQTT broker across multiple user is risky.

One option to avoid someone to access the cmnd/topic is to rename the cmnd prefix with a private key. See command `Prefix1`.
Let's says that each participant rename cmnd to a key made of 16 alphanum characters that he only knows. So only him can send a command to his device.

**Reply by ouinouin** [reply]
i agree that mqtt is not fit for multiple user,  we also look ahead to give additional material for user interrested in setting a local infra, but id like to have a setup relying only on an internet access, i like the idea of the prefix1 trick since its available at no additional developping cost , i ll also try to disable completely the subscription to the CMND topic and make this setting accessible through a tick button on the web interface.
to your knowledge , disabling cmnd subscription would have some adverse effect ?

**Reply by barbudor** [reply]
I don't see any side effect than preventing the control of the device through MQTT.
You would also have to disable the fallback command topic

**Reply by ouinouin** [reply]
i also have to seek to disable the grouptopic i suppose, or there is a setoption to disable the grouptopic ?

**Reply by barbudor** [reply]
no setoption, you will have to handle that too

**Reply by redhat421** [reply]
Consider also using this command: 
[PowerOnState](https://tasmota.github.io/docs/Commands/#poweronstate) 4

That should cause the relay to be enabled and disable further relay control at boot. Even better would be to physically bridge the output on the relay to make it impossible to turn the connected device off.

Additionally, I'd consider having a seperate MQTT instance for each batch of customers, or one per each customer if possible.
I might also patch the MQTT server to prevent it from having the ability to send any messages other than those needed to establish a MQTT connection and get the Telemetry data.

Good luck!
